{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66eb9025-ae99-4ef7-95cc-05bfaac49820",
   "metadata": {},
   "source": [
    "# WS_ch03A.ipynb\n",
    "# WESmith 11/09/22\n",
    "## WS created this notebook to follow along chap 3 code from book \n",
    "# 'Bioinformatics with Python Cookbook' by Tiago Antao\n",
    "### Each recipe will have its own notebook, suffixed by A, B, etc.\n",
    "## see the link below for reference to SeqIO data structures\n",
    "### http://biopython.org/DIST/docs/tutorial/Tutorial.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43246f84-be54-4edd-bacb-b4bbd8212e62",
   "metadata": {},
   "source": [
    "# ACCESSING GENBANK AND MOVING AROUND NCBI DATABASES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f2f2140d-70b9-425c-8ba0-a4f540a540cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "82a02f68-6d09-427f-99e1-9871c4c651e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = 'smiwarsky@gmail.com'  # required"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d9a9811b-7abb-48b4-a0ce-1085702650a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the chloroquinine resistance transporter (CRT) gene in Plasmodium falciparum\n",
    "handle = Entrez.esearch(db='nucleotide', term='CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]')\n",
    "rec_list = Entrez.read(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "7428605d-2b00-443e-afa9-e6feda8b49f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count = 2925\n",
      "RetMax = 20\n",
      "RetStart = 0\n",
      "IdList = ['2301594124', '2301594089', '2262825096', '2262825094', '2262825092', '2262825090', '2262825088', '2262825086', '2262825084', '2262825082', '2262825080', '2262825078', '2262825076', '2262825074', '2262825072', '2262825070', '2262825068', '2262825066', '2262825064', '2262825062']\n",
      "TranslationSet = [{'From': '\"Plasmodium falciparum\"[Organism]', 'To': '\"Plasmodium falciparum\"[Organism]'}]\n",
      "TranslationStack = [{'Term': 'CRT[Gene Name]', 'Field': 'Gene Name', 'Count': '5780', 'Explode': 'N'}, {'Term': '\"Plasmodium falciparum\"[Organism]', 'Field': 'Organism', 'Count': '261968', 'Explode': 'Y'}, 'AND']\n",
      "QueryTranslation = CRT[Gene Name] AND \"Plasmodium falciparum\"[Organism]\n"
     ]
    }
   ],
   "source": [
    "for j, k in rec_list.items():\n",
    "    print('{} = {}'.format(j, k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "31be5cc1-8b23-4283-a895-a76d55a6b408",
   "metadata": {},
   "outputs": [],
   "source": [
    "hdl = Entrez.efetch(db='nucleotide', id=id_list, rettype='gb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "74c371cf-dca7-43fd-89fd-92e2a8c19dfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "_io.TextIOWrapper"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(hdl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fd9b9ed2-93de-42e2-ae73-bce3ebe88ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "recs = list(SeqIO.parse(hdl,'gb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "3e2e3422-9100-46f7-a071-c41964d9da2d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on SeqRecord in module Bio.SeqRecord object:\n",
      "\n",
      "class SeqRecord(builtins.object)\n",
      " |  SeqRecord(seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |  \n",
      " |  A SeqRecord object holds a sequence and information about it.\n",
      " |  \n",
      " |  Main attributes:\n",
      " |   - id          - Identifier such as a locus tag (string)\n",
      " |   - seq         - The sequence itself (Seq object or similar)\n",
      " |  \n",
      " |  Additional attributes:\n",
      " |   - name        - Sequence name, e.g. gene name (string)\n",
      " |   - description - Additional text (string)\n",
      " |   - dbxrefs     - List of database cross references (list of strings)\n",
      " |   - features    - Any (sub)features defined (list of SeqFeature objects)\n",
      " |   - annotations - Further information about the whole sequence (dictionary).\n",
      " |     Most entries are strings, or lists of strings.\n",
      " |   - letter_annotations - Per letter/symbol annotation (restricted\n",
      " |     dictionary). This holds Python sequences (lists, strings\n",
      " |     or tuples) whose length matches that of the sequence.\n",
      " |     A typical use would be to hold a list of integers\n",
      " |     representing sequencing quality scores, or a string\n",
      " |     representing the secondary structure.\n",
      " |  \n",
      " |  You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |  SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |  objects directly (see the __init__ method for further details):\n",
      " |  \n",
      " |  >>> from Bio.Seq import Seq\n",
      " |  >>> from Bio.SeqRecord import SeqRecord\n",
      " |  >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |  ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |  ...                    description=\"toxic membrane protein\")\n",
      " |  >>> print(record)\n",
      " |  ID: YP_025292.1\n",
      " |  Name: HokC\n",
      " |  Description: toxic membrane protein\n",
      " |  Number of features: 0\n",
      " |  Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |  \n",
      " |  If you want to save SeqRecord objects to a sequence file, use Bio.SeqIO\n",
      " |  for this.  For the special case where you want the SeqRecord turned into\n",
      " |  a string in a particular file format there is a format method which uses\n",
      " |  Bio.SeqIO internally:\n",
      " |  \n",
      " |  >>> print(record.format(\"fasta\"))\n",
      " |  >YP_025292.1 toxic membrane protein\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also do things like slicing a SeqRecord, checking its length, etc\n",
      " |  \n",
      " |  >>> len(record)\n",
      " |  44\n",
      " |  >>> edited = record[:10] + record[11:]\n",
      " |  >>> print(edited.seq)\n",
      " |  MKQHKAMIVAIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  >>> print(record.seq)\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      The other sequence can be a SeqRecord object, a Seq object (or\n",
      " |      similar, e.g. a MutableSeq) or a plain Python string. If you add\n",
      " |      a plain string or a Seq (like) object, the new SeqRecord will simply\n",
      " |      have this appended to the existing data. However, any per letter\n",
      " |      annotation will be lost:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = record + \"ACT\"\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNNACT\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |      \n",
      " |      The new record will attempt to combine the annotation, but for any\n",
      " |      ambiguities (e.g. different names) it defaults to omitting that\n",
      " |      annotation.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      \n",
      " |      Now let's cut the plasmid into two pieces, and join them back up the\n",
      " |      other way round (i.e. shift the starting point on this plasmid, have\n",
      " |      a look at the annotated features in the original file to see why this\n",
      " |      particular split point might make sense):\n",
      " |      \n",
      " |      >>> left = plasmid[:3765]\n",
      " |      >>> right = plasmid[3765:]\n",
      " |      >>> new = right + left\n",
      " |      >>> print(\"%s %i\" % (new.id, len(new)))\n",
      " |      pBAD30 4923\n",
      " |      >>> str(new.seq) == str(right.seq + left.seq)\n",
      " |      True\n",
      " |      >>> len(new.features) == len(left.features) + len(right.features)\n",
      " |      True\n",
      " |      \n",
      " |      When we add the left and right SeqRecord objects, their annotation\n",
      " |      is all consistent, so it is all conserved in the new SeqRecord:\n",
      " |      \n",
      " |      >>> new.id == left.id == right.id == plasmid.id\n",
      " |      True\n",
      " |      >>> new.name == left.name == right.name == plasmid.name\n",
      " |      True\n",
      " |      >>> new.description == plasmid.description\n",
      " |      True\n",
      " |      >>> new.annotations == left.annotations == right.annotations\n",
      " |      True\n",
      " |      >>> new.letter_annotations == plasmid.letter_annotations\n",
      " |      True\n",
      " |      >>> new.dbxrefs == left.dbxrefs == right.dbxrefs\n",
      " |      True\n",
      " |      \n",
      " |      However, we should point out that when we sliced the SeqRecord,\n",
      " |      any annotations dictionary or dbxrefs list entries were lost.\n",
      " |      You can explicitly copy them like this:\n",
      " |      \n",
      " |      >>> new.annotations = plasmid.annotations.copy()\n",
      " |      >>> new.dbxrefs = plasmid.dbxrefs[:]\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Boolean value of an instance of this class (True).\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a SeqRecord always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, a Seq object will evaluate to False if it\n",
      " |      has a zero length sequence.\n",
      " |      \n",
      " |      WARNING: The SeqRecord may in future evaluate to False when its\n",
      " |      sequence is of zero length (in order to better match the Seq\n",
      " |      object behaviour)!\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implement the 'in' keyword, searches the sequence.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> \"GAATTC\" in record\n",
      " |      False\n",
      " |      >>> \"AAA\" in record\n",
      " |      True\n",
      " |      \n",
      " |      This essentially acts as a proxy for using \"in\" on the sequence:\n",
      " |      \n",
      " |      >>> \"GAATTC\" in record.seq\n",
      " |      False\n",
      " |      >>> \"AAA\" in record.seq\n",
      " |      True\n",
      " |      \n",
      " |      Note that you can also use Seq objects as the query,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> Seq(\"AAA\") in record\n",
      " |      True\n",
      " |      \n",
      " |      See also the Seq object's __contains__ method.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Define the equal-to operand (not implemented).\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      This method supports the Python format() function and f-strings.\n",
      " |      The format_spec should be a lower case string supported by\n",
      " |      Bio.SeqIO as a text output file format. Requesting a binary file\n",
      " |      format raises a ValueError. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      ...\n",
      " |      >>> format(record, \"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(f\"Here is {record.id} in FASTA format:\\n{record:fasta}\")\n",
      " |      Here is YP_025292.1 in FASTA format:\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      See also the SeqRecord's format() method.\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Define the greater-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a sub-sequence or an individual letter.\n",
      " |      \n",
      " |      Slicing, e.g. my_record[5:10], returns a new SeqRecord for\n",
      " |      that sub-sequence with some annotation preserved as follows:\n",
      " |      \n",
      " |      * The name, id and description are kept as-is.\n",
      " |      * Any per-letter-annotations are sliced to match the requested\n",
      " |        sub-sequence.\n",
      " |      * Unless a stride is used, all those features which fall fully\n",
      " |        within the subsequence are included (with their locations\n",
      " |        adjusted accordingly). If you want to preserve any truncated\n",
      " |        features (e.g. GenBank/EMBL source features), you must\n",
      " |        explicitly add them to the new SeqRecord yourself.\n",
      " |      * With the exception of any molecule type, the annotations\n",
      " |        dictionary and the dbxrefs list are not used for the new\n",
      " |        SeqRecord, as in general they may not apply to the\n",
      " |        subsequence. If you want to preserve them, you must explicitly\n",
      " |        copy them to the new SeqRecord yourself.\n",
      " |      \n",
      " |      Using an integer index, e.g. my_record[5] is shorthand for\n",
      " |      extracting that letter from the sequence, my_record.seq[5].\n",
      " |      \n",
      " |      For example, consider this short protein and its secondary\n",
      " |      structure as encoded by the PDB (e.g. H for alpha helices),\n",
      " |      plus a simple feature for its histidine self phosphorylation\n",
      " |      site:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      " |      >>> rec = SeqRecord(Seq(\"MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLAT\"\n",
      " |      ...                     \"EMMSEQDGYLAESINKDIEECNAIIEQFIDYLR\"),\n",
      " |      ...                 id=\"1JOY\", name=\"EnvZ\",\n",
      " |      ...                 description=\"Homodimeric domain of EnvZ from E. coli\")\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"] = \"  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \"\n",
      " |      >>> rec.features.append(SeqFeature(FeatureLocation(20, 21),\n",
      " |      ...                     type = \"Site\"))\n",
      " |      \n",
      " |      Now let's have a quick look at the full record,\n",
      " |      \n",
      " |      >>> print(rec)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR')\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"]\n",
      " |      '  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  '\n",
      " |      >>> print(rec.features[0].location)\n",
      " |      [20:21]\n",
      " |      \n",
      " |      Now let's take a sub sequence, here chosen as the first (fractured)\n",
      " |      alpha helix which includes the histidine phosphorylation site:\n",
      " |      \n",
      " |      >>> sub = rec[11:41]\n",
      " |      >>> print(sub)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('RTLLMAGVSHDLRTPLTRIRLATEMMSEQD')\n",
      " |      >>> sub.letter_annotations[\"secondary_structure\"]\n",
      " |      'HHHHHTTTHHHHHHHHHHHHHHHHHHHHHH'\n",
      " |      >>> print(sub.features[0].location)\n",
      " |      [9:10]\n",
      " |      \n",
      " |      You can also of course omit the start or end values, for\n",
      " |      example to get the first ten letters only:\n",
      " |      \n",
      " |      >>> print(rec[:10])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLAD')\n",
      " |      \n",
      " |      Or for the last ten letters:\n",
      " |      \n",
      " |      >>> print(rec[-10:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('IIEQFIDYLR')\n",
      " |      \n",
      " |      If you omit both, then you get a copy of the original record (although\n",
      " |      lacking the annotations and dbxrefs):\n",
      " |      \n",
      " |      >>> print(rec[:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR')\n",
      " |      \n",
      " |      Finally, indexing with a simple integer is shorthand for pulling out\n",
      " |      that letter from the sequence directly:\n",
      " |      \n",
      " |      >>> rec[5]\n",
      " |      'K'\n",
      " |      >>> rec.seq[5]\n",
      " |      'K'\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Define the greater-than operand (not implemented).\n",
      " |  \n",
      " |  __init__(self, seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |      Create a SeqRecord.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - seq         - Sequence, required (Seq or MutableSeq)\n",
      " |       - id          - Sequence identifier, recommended (string)\n",
      " |       - name        - Sequence name, optional (string)\n",
      " |       - description - Sequence description, optional (string)\n",
      " |       - dbxrefs     - Database cross references, optional (list of strings)\n",
      " |       - features    - Any (sub)features, optional (list of SeqFeature objects)\n",
      " |       - annotations - Dictionary of annotations for the whole sequence\n",
      " |       - letter_annotations - Dictionary of per-letter-annotations, values\n",
      " |         should be strings, list or tuples of the same length as the full\n",
      " |         sequence.\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |      objects directly.\n",
      " |      \n",
      " |      Note that while an id is optional, we strongly recommend you supply a\n",
      " |      unique id string for each record.  This is especially important\n",
      " |      if you wish to write your sequences to a file.\n",
      " |      \n",
      " |      You can create a 'blank' SeqRecord object, and then populate the\n",
      " |      attributes later.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the letters in the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a protein FASTA file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/loveliesbleeding.pro\", \"fasta\")\n",
      " |      >>> for amino in record:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      This is just a shortcut for iterating over the sequence directly:\n",
      " |      \n",
      " |      >>> for amino in record.seq:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      Note that this does not facilitate iteration together with any\n",
      " |      per-letter-annotation.  However, you can achieve that using the\n",
      " |      python zip function on the record (or its sequence) and the relevant\n",
      " |      per-letter-annotation:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> rec = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(rec.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> for nuc, qual in zip(rec, rec.letter_annotations[\"solexa_quality\"]):\n",
      " |      ...     if qual > 35:\n",
      " |      ...         print(\"%s %i\" % (nuc, qual))\n",
      " |      A 40\n",
      " |      C 39\n",
      " |      G 38\n",
      " |      T 37\n",
      " |      A 36\n",
      " |      \n",
      " |      You may agree that using zip(rec.seq, ...) is more explicit than using\n",
      " |      zip(rec, ...) as shown above.\n",
      " |  \n",
      " |  __le___(self, other)\n",
      " |      Define the less-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a FASTA nucleotide file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> len(record)\n",
      " |      309\n",
      " |      >>> len(record.seq)\n",
      " |      309\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Define the less-than operand (not implemented).\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Define the not-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add another sequence or string to this sequence (from the left).\n",
      " |      \n",
      " |      This method handles adding a Seq object (or similar, e.g. MutableSeq)\n",
      " |      or a plain Python string (on the left) to a SeqRecord (on the right).\n",
      " |      See the __add__ method for more details, but for example:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = \"ACT\" + record\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a concise summary of the record for debugging (string).\n",
      " |      \n",
      " |      The python built in function repr works by calling the object's ___repr__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rec = SeqRecord(Seq(\"MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKAT\"\n",
      " |      ...                     \"GEMKEQTEWHRVVLFGKLAEVASEYLRKGSQVYIEGQLRTRKWTDQ\"\n",
      " |      ...                     \"SGQDRYTTEVVVNVGGTMQMLGGRQGGGAPAGGNIGGGQPQGGWGQ\"\n",
      " |      ...                     \"PQQPQGGNQFSGGAQSRPQQSAPAAPSNEPPMDFDDDIPF\"),\n",
      " |      ...                 id=\"NP_418483.1\", name=\"b4059\",\n",
      " |      ...                 description=\"ssDNA-binding protein\",\n",
      " |      ...                 dbxrefs=[\"ASAP:13298\", \"GI:16131885\", \"GeneID:948570\"])\n",
      " |      >>> print(repr(rec))\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF'), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      At the python prompt you can also use this shorthand:\n",
      " |      \n",
      " |      >>> rec\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF'), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      Note that long sequences are shown truncated. Also note that any\n",
      " |      annotations, letter_annotations and features are not shown (as they\n",
      " |      would lead to a very long string).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a human readable summary of the record and its annotation (string).\n",
      " |      \n",
      " |      The python built in function str works by calling the object's ___str__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein, small\")\n",
      " |      >>> print(str(record))\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      \n",
      " |      In this example you don't actually need to call str explicity, as the\n",
      " |      print command does this automatically:\n",
      " |      \n",
      " |      >>> print(record)\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      \n",
      " |      Note that long sequences are shown truncated.\n",
      " |  \n",
      " |  format(self, format)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.SeqIO, which is used to turn the SeqRecord into a\n",
      " |      string.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      >>> record.format(\"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      The Python print function automatically appends a new line, meaning\n",
      " |      in this example a blank line is shown.  If you look at the string\n",
      " |      representation you can see there is a trailing new line (shown as\n",
      " |      slash n) which is important when writing to a file or if\n",
      " |      concatenating multiple sequence strings together.\n",
      " |      \n",
      " |      Note that this method will NOT work on every possible file format\n",
      " |      supported by Bio.SeqIO (e.g. some are for multiple sequences only,\n",
      " |      and binary formats are not supported).\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Return a copy of the record with a lower case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/aster.pro\", \"fasta\")\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      GGHVNPAVTFGAFVGGNITLLRGIVYIIAQLLGSTVACLLLKFVTNDMAVGVFSLSAGVG\n",
      " |      VTNALVFEIVMTFGLVYTVYATAIDPKKGSLGTIAPIAIGFIVGANI\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(record.lower().format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      gghvnpavtfgafvggnitllrgivyiiaqllgstvaclllkfvtndmavgvfslsagvg\n",
      " |      vtnalvfeivmtfglvytvyataidpkkgslgtiapiaigfivgani\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      To take a more annotation rich example,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> old = SeqIO.read(\"EMBL/TRBG361.embl\", \"embl\")\n",
      " |      >>> len(old.features)\n",
      " |      3\n",
      " |      >>> new = old.lower()\n",
      " |      >>> len(old.features) == len(new.features)\n",
      " |      True\n",
      " |      >>> old.annotations[\"organism\"] == new.annotations[\"organism\"]\n",
      " |      True\n",
      " |      >>> old.dbxrefs == new.dbxrefs\n",
      " |      True\n",
      " |  \n",
      " |  reverse_complement(self, id=False, name=False, description=False, features=True, annotations=False, letter_annotations=True, dbxrefs=False)\n",
      " |      Return new SeqRecord with reverse complement sequence.\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the reversed sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or True to keep that of the parent, or False to\n",
      " |      omit them. The default is to keep the original features (with the\n",
      " |      strand and locations adjusted).\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent,\n",
      " |      or False to omit them. The default is to keep the original\n",
      " |      annotations (with the letter annotations reversed).\n",
      " |      \n",
      " |      To show what happens to the pre-letter annotations, consider an\n",
      " |      example Solexa variant FASTQ file with a single entry, which we'll\n",
      " |      read in as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Now take the reverse complement, here we explicitly give a new\n",
      " |      identifier (the old identifier with a suffix):\n",
      " |      \n",
      " |      >>> rc_record = record.reverse_complement(id=record.id + \"_rc\")\n",
      " |      >>> print(\"%s %s\" % (rc_record.id, rc_record.seq))\n",
      " |      slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT\n",
      " |      \n",
      " |      Notice that the per-letter-annotations have also been reversed,\n",
      " |      although this may not be appropriate for all cases.\n",
      " |      \n",
      " |      >>> print(rc_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]\n",
      " |      \n",
      " |      Now for the features, we need a different example. Parsing a GenBank\n",
      " |      file is probably the easiest way to get an nice example with features\n",
      " |      in it...\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      >>> plasmid.seq\n",
      " |      Seq('GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG')\n",
      " |      >>> len(plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Now, let's take the reverse complement of this whole plasmid:\n",
      " |      \n",
      " |      >>> rc_plasmid = plasmid.reverse_complement(id=plasmid.id+\"_rc\")\n",
      " |      >>> print(\"%s %i\" % (rc_plasmid.id, len(rc_plasmid)))\n",
      " |      pBAD30_rc 4923\n",
      " |      >>> rc_plasmid.seq\n",
      " |      Seq('CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC')\n",
      " |      >>> len(rc_plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Let's compare the first CDS feature - it has gone from being the\n",
      " |      second feature (index 1) to the second last feature (index -2), its\n",
      " |      strand has changed, and the location switched round.\n",
      " |      \n",
      " |      >>> print(plasmid.features[1])\n",
      " |      type: CDS\n",
      " |      location: [1081:1960](-)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rc_plasmid.features[-2])\n",
      " |      type: CDS\n",
      " |      location: [2963:3842](+)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      You can check this new location, based on the length of the plasmid:\n",
      " |      \n",
      " |      >>> len(plasmid) - 1081\n",
      " |      3842\n",
      " |      >>> len(plasmid) - 1960\n",
      " |      2963\n",
      " |      \n",
      " |      Note that if the SeqFeature annotation includes any strand specific\n",
      " |      information (e.g. base changes for a SNP), this information is not\n",
      " |      amended, and would need correction after the reverse complement.\n",
      " |      \n",
      " |      Note trying to reverse complement a protein SeqRecord raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> protein_rec = SeqRecord(Seq(\"MAIVMGR\"), id=\"Test\",\n",
      " |      ...                         annotations={\"molecule_type\": \"protein\"})\n",
      " |      >>> protein_rec.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |      \n",
      " |      If you have RNA without any U bases, it must be annotated as RNA\n",
      " |      otherwise it will be treated as DNA by default with A mapped to T:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rna1 = SeqRecord(Seq(\"ACG\"), id=\"Test\")\n",
      " |      >>> rna2 = SeqRecord(Seq(\"ACG\"), id=\"Test\", annotations={\"molecule_type\": \"RNA\"})\n",
      " |      >>> print(rna1.reverse_complement(id=\"RC\", description=\"unk\").format(\"fasta\"))\n",
      " |      >RC unk\n",
      " |      CGT\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rna2.reverse_complement(id=\"RC\", description=\"RNA\").format(\"fasta\"))\n",
      " |      >RC RNA\n",
      " |      CGU\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Also note you can reverse complement a SeqRecord using a MutableSeq:\n",
      " |      \n",
      " |      >>> from Bio.Seq import MutableSeq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rec = SeqRecord(MutableSeq(\"ACGT\"), id=\"Test\")\n",
      " |      >>> rec.seq[0] = \"T\"\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      Test TCGT\n",
      " |      >>> rc = rec.reverse_complement(id=True)\n",
      " |      >>> print(\"%s %s\" % (rc.id, rc.seq))\n",
      " |      Test ACGA\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap=None, id=False, name=False, description=False, features=False, annotations=False, letter_annotations=False, dbxrefs=False)\n",
      " |      Return new SeqRecord with translated sequence.\n",
      " |      \n",
      " |      This calls the record's .seq.translate() method (which describes\n",
      " |      the translation related arguments, like table for the genetic code),\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the translated sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or False (default) to omit them.\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent\n",
      " |      (annotations only), or False (default) to omit them.\n",
      " |      \n",
      " |      e.g. Loading a FASTA gene and translating it,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> gene_record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> print(gene_record.format(\"fasta\"))\n",
      " |      >gi|3176602|gb|U78617.1|LOU78617 Lathyrus odoratus phytochrome A (PHYA) gene, partial cds\n",
      " |      CAGGCTGCGCGGTTTCTATTTATGAAGAACAAGGTCCGTATGATAGTTGATTGTCATGCA\n",
      " |      AAACATGTGAAGGTTCTTCAAGACGAAAAACTCCCATTTGATTTGACTCTGTGCGGTTCG\n",
      " |      ACCTTAAGAGCTCCACATAGTTGCCATTTGCAGTACATGGCTAACATGGATTCAATTGCT\n",
      " |      TCATTGGTTATGGCAGTGGTCGTCAATGACAGCGATGAAGATGGAGATAGCCGTGACGCA\n",
      " |      GTTCTACCACAAAAGAAAAAGAGACTTTGGGGTTTGGTAGTTTGTCATAACACTACTCCG\n",
      " |      AGGTTTGTT\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      And now translating the record, specifying the new ID and description:\n",
      " |      \n",
      " |      >>> protein_record = gene_record.translate(table=11,\n",
      " |      ...                                        id=\"phya\",\n",
      " |      ...                                        description=\"translation\")\n",
      " |      >>> print(protein_record.format(\"fasta\"))\n",
      " |      >phya translation\n",
      " |      QAARFLFMKNKVRMIVDCHAKHVKVLQDEKLPFDLTLCGSTLRAPHSCHLQYMANMDSIA\n",
      " |      SLVMAVVVNDSDEDGDSRDAVLPQKKKRLWGLVVCHNTTPRFV\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Return a copy of the record with an upper case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"acgtACGT\"), id=\"Test\",\n",
      " |      ...                    description = \"Made up for this example\")\n",
      " |      >>> record.letter_annotations[\"phred_quality\"] = [1, 2, 3, 4, 5, 6, 7, 8]\n",
      " |      >>> print(record.upper().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      ACGTACGT\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Naturally, there is a matching lower method:\n",
      " |      \n",
      " |      >>> print(record.lower().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      acgtacgt\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  letter_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |      \n",
      " |      For example, this can hold quality scores used in FASTQ or QUAL files.\n",
      " |      Consider this example using Bio.SeqIO to read in an example Solexa\n",
      " |      variant FASTQ file as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      The letter_annotations get sliced automatically if you slice the\n",
      " |      parent SeqRecord, for example taking the last ten bases:\n",
      " |      \n",
      " |      >>> sub_record = record[-10:]\n",
      " |      >>> print(\"%s %s\" % (sub_record.id, sub_record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTNNNNNN\n",
      " |      >>> print(sub_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Any python sequence (i.e. list, tuple or string) can be recorded in\n",
      " |      the SeqRecord's letter_annotations dictionary as long as the length\n",
      " |      matches that of the SeqRecord's sequence.  e.g.\n",
      " |      \n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      1\n",
      " |      >>> sub_record.letter_annotations[\"dummy\"] = \"abcdefghij\"\n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      2\n",
      " |      \n",
      " |      You can delete entries from the letter_annotations dictionary as usual:\n",
      " |      \n",
      " |      >>> del sub_record.letter_annotations[\"solexa_quality\"]\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {'dummy': 'abcdefghij'}\n",
      " |      \n",
      " |      You can completely clear the dictionary easily as follows:\n",
      " |      \n",
      " |      >>> sub_record.letter_annotations = {}\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {}\n",
      " |      \n",
      " |      Note that if replacing the record's sequence with a sequence of a\n",
      " |      different length you must first clear the letter_annotations dict.\n",
      " |  \n",
      " |  seq\n",
      " |      The sequence itself, as a Seq or MutableSeq object.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(recs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "bd5e6dd3-779f-4dde-9773-c0abf6471732",
   "metadata": {},
   "outputs": [],
   "source": [
    "def attrs(obj, skip=True, token='__'):  # WS convenience function\n",
    "    attr = ['OBJECT TYPE: {}'.format(type(obj))]\n",
    "    for k in dir(obj):\n",
    "        if skip and k.__contains__(token): continue\n",
    "        attr.append(k)\n",
    "    return attr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "b2ff678f-650e-48c5-8cb5-85b092c5d347",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"OBJECT TYPE: <class 'Bio.SeqRecord.SeqRecord'>\",\n",
       " '_per_letter_annotations',\n",
       " '_seq',\n",
       " '_set_per_letter_annotations',\n",
       " '_set_seq',\n",
       " 'annotations',\n",
       " 'dbxrefs',\n",
       " 'description',\n",
       " 'features',\n",
       " 'format',\n",
       " 'id',\n",
       " 'letter_annotations',\n",
       " 'lower',\n",
       " 'name',\n",
       " 'reverse_complement',\n",
       " 'seq',\n",
       " 'translate',\n",
       " 'upper']"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs(recs[0], skip=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "d0d13f54-2ac0-4c44-a388-d2329e4e451c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"OBJECT TYPE: <class 'Bio.SeqRecord.SeqRecord'>\",\n",
       " '_per_letter_annotations',\n",
       " '_seq',\n",
       " '_set_per_letter_annotations',\n",
       " '_set_seq',\n",
       " 'annotations',\n",
       " 'dbxrefs',\n",
       " 'description',\n",
       " 'features',\n",
       " 'format',\n",
       " 'id',\n",
       " 'letter_annotations',\n",
       " 'lower',\n",
       " 'name',\n",
       " 'reverse_complement',\n",
       " 'seq',\n",
       " 'translate',\n",
       " 'upper']"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs(recs[0].reverse_complement())  # WS exploring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "8103675b-5483-487a-855a-3358c5e01778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'molecule_type': 'DNA',\n",
       " 'topology': 'linear',\n",
       " 'data_file_division': 'INV',\n",
       " 'date': '19-SEP-2022',\n",
       " 'accessions': ['MZ054304'],\n",
       " 'sequence_version': 1,\n",
       " 'keywords': [''],\n",
       " 'source': 'Plasmodium falciparum 3D7',\n",
       " 'organism': 'Plasmodium falciparum 3D7',\n",
       " 'taxonomy': ['Eukaryota',\n",
       "  'Sar',\n",
       "  'Alveolata',\n",
       "  'Apicomplexa',\n",
       "  'Aconoidasida',\n",
       "  'Haemosporida',\n",
       "  'Plasmodiidae',\n",
       "  'Plasmodium',\n",
       "  'Plasmodium (Laverania)'],\n",
       " 'references': [Reference(title='Direct Submission', ...)],\n",
       " 'structured_comment': OrderedDict([('Assembly-Data',\n",
       "               OrderedDict([('Sequencing Technology',\n",
       "                             'Sanger dideoxy sequencing')]))])}"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[0].annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "c4cf5f43-9b82-4cc9-8149-1b393b5fbb39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Plasmodium falciparum 3D7 isolate CRT33 chloroquine resistance transporter (crt) gene, partial cds'"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[0].description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "45bcfef2-4273-43e0-91f9-831ce4b4c373",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(209), strand=1), type='source'),\n",
       " SeqFeature(FeatureLocation(BeforePosition(0), AfterPosition(209), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(BeforePosition(0), AfterPosition(209), strand=1), type='mRNA'),\n",
       " SeqFeature(FeatureLocation(BeforePosition(0), AfterPosition(209), strand=1), type='CDS')]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[0].features  # WS exploring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "4a480be4-e942-4f65-b58d-54cf924b0281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"OBJECT TYPE: <class 'Bio.SeqFeature.SeqFeature'>\",\n",
       " '_flip',\n",
       " '_get_location_operator',\n",
       " '_get_ref',\n",
       " '_get_ref_db',\n",
       " '_get_strand',\n",
       " '_set_location_operator',\n",
       " '_set_ref',\n",
       " '_set_ref_db',\n",
       " '_set_strand',\n",
       " '_shift',\n",
       " 'extract',\n",
       " 'id',\n",
       " 'location',\n",
       " 'location_operator',\n",
       " 'qualifiers',\n",
       " 'ref',\n",
       " 'ref_db',\n",
       " 'strand',\n",
       " 'translate',\n",
       " 'type']"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs(recs[0].features[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "c008e154-92d5-4ff8-97b9-ace460c57882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"OBJECT TYPE: <class 'collections.OrderedDict'>\",\n",
       " 'clear',\n",
       " 'copy',\n",
       " 'fromkeys',\n",
       " 'get',\n",
       " 'items',\n",
       " 'keys',\n",
       " 'move_to_end',\n",
       " 'pop',\n",
       " 'popitem',\n",
       " 'setdefault',\n",
       " 'update',\n",
       " 'values']"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs(recs[0].features[0].qualifiers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "49b63e45-9222-45a5-acbd-27973e35b005",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MZ054304\n",
      "Plasmodium falciparum 3D7 isolate CRT33 chloroquine resistance transporter (crt) gene, partial cds\n",
      "\n",
      "MZ054303\n",
      "Plasmodium falciparum 3D7 isolate CRT4 chloroquine resistance transporter (crt) gene, partial cds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for jj in range(2):\n",
    "    print('{}\\n{}\\n'.format(recs[jj].name, recs[jj].description))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "99ec6037-692c-4d13-8ef8-3cef3802a75f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OrderedDict([('organism', ['Plasmodium falciparum 3D7']), ('mol_type', ['genomic DNA']), ('isolate', ['CRT33']), ('host', ['Homo sapiens']), ('db_xref', ['taxon:36329']), ('chromosome', ['7']), ('country', ['India: Nuapada District Hq Hospital']), ('collection_date', ['14-Jul-2019']), ('collected_by', ['Ramakanta Rana']), ('PCR_primers', ['fwd_name: pfcrt-f, fwd_seq: ggctcacgtttaggtgga, rev_name: pfcrt-r, rev_seq: tgaatttccctttttatttccaaa'])])\n",
      "\n",
      "OrderedDict([('gene', ['crt']), ('note', ['pfcrt'])])\n",
      "\n",
      "OrderedDict([('gene', ['crt']), ('product', ['chloroquine resistance transporter'])])\n",
      "\n",
      "OrderedDict([('gene', ['crt']), ('codon_start', ['1']), ('product', ['chloroquine resistance transporter']), ('protein_id', ['UWU44816.1']), ('translation', ['LIFKEIKDNIFIYILSIIYLSVCVMNKIFAKRTLNKIGNYSFVTSETHNFICMIMFFIVYSLFGNKKGNS'])])\n"
     ]
    }
   ],
   "source": [
    "# WS exploring\n",
    "for k in recs[0].features:\n",
    "    print('\\n{}'.format(k.qualifiers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "130ec857-e372-46cf-bd14-49c648f1e113",
   "metadata": {},
   "outputs": [],
   "source": [
    "# WS created function, not done in text\n",
    "def get_info(rec):\n",
    "    for feature in rec.features:\n",
    "        #print('FEATURE TYPE: {}\\n'.format(feature.type))\n",
    "        if feature.type == 'gene':\n",
    "            print('\\nPROCESSED: GENE TYPE {}'.format(feature.qualifiers['gene']))\n",
    "        elif feature.type == 'exon':\n",
    "            loc = feature.location\n",
    "            print('\\nPROCESSED: EXON TYPE {}'.format(loc.start, loc.end, loc.strand))\n",
    "        else:\n",
    "            print('\\nNOT PROCESSED: {}'.format(feature))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "eb1fcf97-2695-4375-815a-cb186a441f4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NOT PROCESSED: type: source\n",
      "location: [0:209](+)\n",
      "qualifiers:\n",
      "    Key: PCR_primers, Value: ['fwd_name: pfcrt-f, fwd_seq: ggctcacgtttaggtgga, rev_name: pfcrt-r, rev_seq: tgaatttccctttttatttccaaa']\n",
      "    Key: chromosome, Value: ['7']\n",
      "    Key: collected_by, Value: ['Ramakanta Rana']\n",
      "    Key: collection_date, Value: ['14-Jul-2019']\n",
      "    Key: country, Value: ['India: Nuapada District Hq Hospital']\n",
      "    Key: db_xref, Value: ['taxon:36329']\n",
      "    Key: host, Value: ['Homo sapiens']\n",
      "    Key: isolate, Value: ['CRT33']\n",
      "    Key: mol_type, Value: ['genomic DNA']\n",
      "    Key: organism, Value: ['Plasmodium falciparum 3D7']\n",
      "\n",
      "\n",
      "PROCESSED: GENE TYPE ['crt']\n",
      "\n",
      "NOT PROCESSED: type: mRNA\n",
      "location: [<0:>209](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['crt']\n",
      "    Key: product, Value: ['chloroquine resistance transporter']\n",
      "\n",
      "\n",
      "NOT PROCESSED: type: CDS\n",
      "location: [<0:>209](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: gene, Value: ['crt']\n",
      "    Key: product, Value: ['chloroquine resistance transporter']\n",
      "    Key: protein_id, Value: ['UWU44816.1']\n",
      "    Key: translation, Value: ['LIFKEIKDNIFIYILSIIYLSVCVMNKIFAKRTLNKIGNYSFVTSETHNFICMIMFFIVYSLFGNKKGNS']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "get_info(recs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "7d91a7b9-04c9-489e-83f8-b0206a29c859",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOLECULE_TYPE = DNA\n",
      "TOPOLOGY = linear\n",
      "DATA_FILE_DIVISION = INV\n",
      "DATE = 06-JUL-2022\n",
      "ACCESSIONS = ['OM964461']\n",
      "SEQUENCE_VERSION = 1\n",
      "KEYWORDS = ['']\n",
      "SOURCE = Plasmodium falciparum (malaria parasite P. falciparum)\n",
      "ORGANISM = Plasmodium falciparum\n",
      "TAXONOMY = ['Eukaryota', 'Sar', 'Alveolata', 'Apicomplexa', 'Aconoidasida', 'Haemosporida', 'Plasmodiidae', 'Plasmodium', 'Plasmodium (Laverania)']\n",
      "REFERENCES = [Reference(title='Emergence and expansion of piperaquine resistance mediated by the PfCRTC350R mutation and plasmepsin gene amplifications hinder dihydroartemisinin-piperaquine efficacy in the Guiana Shield', ...), Reference(title='Direct Submission', ...)]\n",
      "STRUCTURED_COMMENT = OrderedDict([('Assembly-Data', OrderedDict([('Sequencing Technology', 'Sanger dideoxy sequencing')]))])\n"
     ]
    }
   ],
   "source": [
    "for name, value in recs[10].annotations.items():\n",
    "    print('{} = {}'.format(name.upper(), value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "9ed55ca2-10e5-46fd-bf28-fe74a8b67329",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Bio.Seq.Seq"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(recs[1].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "3d112374-61d7-416c-b5df-42d8efaf9287",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTTATTTTTAAAGAGATTAAGGATAATATTTTTATTTATATTTTAAGTATTATTTATTTAAGTGTATGTGTAATGAATAAAATTTTTGCTAAAAGAACTTTAAACAAAATTGGTAACTATAGTTTTGTAACATCCGAAACTCACAACTTTATTTGTATGATTATGTTCTTTATTGTTTATTCCTTATTTGGAAATAAAAAGGGAAATTCA\n"
     ]
    }
   ],
   "source": [
    "print(recs[1].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "b9ac4e4f-c598-40ed-9c28-7a864628b1e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UGAAUUUCCCUUUUUAUUUCCAAAUAAGGAAUAAACAAUAAAGAACAUAAUCAUACAAAUAAAGUUGUGAGUUUCGGAUGUUACAAAACUAUAGUUACCAAUUUUGUUUAAAGUUCUUUUAGCAAAAAUUUUAUUCAUUACACAUACACUUAAAUAAAUAAUACUUAAAAUAUAAAUAAAAAUAUUAUCCUUAAUCUCUUUAAAAAUAAG\n"
     ]
    }
   ],
   "source": [
    "print(recs[1].seq.reverse_complement_rna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b8e6ef5a-fa13-49e8-8dc8-cec9f22f1dbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 4\n",
      "Seq('TGAATTTCCCTTTTTATTTCCAAATAAGGAATAAACAATAAAGAACATAATCAT...AAG')\n"
     ]
    }
   ],
   "source": [
    "print(recs[1].reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "4c8f381a-b409-4f48-988d-a0a8bfb89438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cttatttttaaagagattaaggataatatttttatttatattttaagtattatttatttaagtgtatgtgtaatgaataaaatttttgctaaaagaactttaaacaaaattggtaactatagttttgtaacatccgaaactcacaactttatttgtatgattatgttctttattgtttattccttatttggaaataaaaagggaaattca\n"
     ]
    }
   ],
   "source": [
    "print(recs[1].seq.lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "03fddca0-e1ef-41f0-8546-5d81a058dad6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGAATTTCCCTTTTTATTTCCAAATAAGGAATAAACAATAAAGAACATAATCATACAAATAAAGTTGTGAGTTTCGGATGTTACAAAACTATAGTTACCAATTTTGTTTAAAGTTCTTTTAGCAAAAATTTTATTCATTACACATACACTTAAATAAATAATACTTAAAATATAAATAAAAATATTATCCTTAATCTCTTTAAAAATAAG\n"
     ]
    }
   ],
   "source": [
    "print(recs[1].seq.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "29b2a542-2d88-4eb8-8f0c-508b2aa2a9a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('SCIQSPALAIAYYFKFLAVRIK')"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs[2].seq.translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "14b11725-25ca-4645-bd07-38fcec75db89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "ID:   MZ054304.1\n",
      "DESC: Plasmodium falciparum 3D7 isolate CRT33 chloroquine resistance transporter (crt) gene, partial cds\n",
      "SEQ:  CTTATTTTTAAAGAGATTAAGGATAATATTTTTATTTATATTTTAAGTATTATTTATTTAAGTGTATGTGTAATGAATAAAATTTTTGCTAAAAGAACTTTAAACAAAATTGGTAACTATAGTTTTGTAACATCCGAAACTCACAACTTTATTTGTATGATTATGTTCTTTATTGTTTATTCCTTATTTGGAAATAAAAAGGGAAATTC\n",
      "TRANSLATE: LIFKEIKDNIFIYILSIIYLSVCVMNKIFAKRTLNKIGNYSFVTSETHNFICMIMFFIVYSLFGNKKGN\n",
      "PUBMED ID: \n",
      "\n",
      "ID:   MZ054303.1\n",
      "DESC: Plasmodium falciparum 3D7 isolate CRT4 chloroquine resistance transporter (crt) gene, partial cds\n",
      "SEQ:  CTTATTTTTAAAGAGATTAAGGATAATATTTTTATTTATATTTTAAGTATTATTTATTTAAGTGTATGTGTAATGAATAAAATTTTTGCTAAAAGAACTTTAAACAAAATTGGTAACTATAGTTTTGTAACATCCGAAACTCACAACTTTATTTGTATGATTATGTTCTTTATTGTTTATTCCTTATTTGGAAATAAAAAGGGAAATTCA\n",
      "TRANSLATE: LIFKEIKDNIFIYILSIIYLSVCVMNKIFAKRTLNKIGNYSFVTSETHNFICMIMFFIVYSLFGNKKGNS\n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964469.1\n",
      "DESC: Plasmodium falciparum isolate SPK77 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAAGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQSPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964468.1\n",
      "DESC: Plasmodium falciparum isolate SPK66 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAAGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQSPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964467.1\n",
      "DESC: Plasmodium falciparum isolate FAF006 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAAGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQSPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964466.1\n",
      "DESC: Plasmodium falciparum isolate SPK83 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964465.1\n",
      "DESC: Plasmodium falciparum isolate SPK81 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964464.1\n",
      "DESC: Plasmodium falciparum isolate SPK80 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964463.1\n",
      "DESC: Plasmodium falciparum isolate SPK76 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964462.1\n",
      "DESC: Plasmodium falciparum isolate FAF007 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964461.1\n",
      "DESC: Plasmodium falciparum isolate FAF008 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964460.1\n",
      "DESC: Plasmodium falciparum isolate FAF010 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964459.1\n",
      "DESC: Plasmodium falciparum isolate SPK75 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964458.1\n",
      "DESC: Plasmodium falciparum isolate SPK12 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964457.1\n",
      "DESC: Plasmodium falciparum isolate SPK13 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964456.1\n",
      "DESC: Plasmodium falciparum isolate SPK14 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964455.1\n",
      "DESC: Plasmodium falciparum isolate SPK17 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964454.1\n",
      "DESC: Plasmodium falciparum isolate SPK18 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964453.1\n",
      "DESC: Plasmodium falciparum isolate SPK74 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n",
      "\n",
      "ID:   OM964452.1\n",
      "DESC: Plasmodium falciparum isolate SPK19 chloroquine resistance transporter (CRT) gene, partial cds\n",
      "SEQ:  AGTTGTATACAAGGTCCAGCATTAGCAATTGCTTATTACTTTAAATTCTTAGCCGTAAGAATTAAA\n",
      "TRANSLATE: SCIQGPALAIAYYFKFLAVRIK\n",
      "PUBMED ID: \n",
      "PUBMED ID: \n"
     ]
    }
   ],
   "source": [
    "for k in recs:\n",
    "    print('\\nID:   {}\\nDESC: {}\\nSEQ:  {}\\nTRANSLATE: {}'.\\\n",
    "          format(k.id, k.description, k.seq, k.seq.translate()))\n",
    "    for ref in k.annotations['references']:\n",
    "        print('PUBMED ID: {}'.format(ref.pubmed_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "d4dcc49b-a875-4bfe-bc0e-64940c3d00df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the above had no pubmed IDs, so didn't take this further: see this recipe in the book"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a168bd-22cf-4cb2-a7ad-8fcb6d12ea81",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
